npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Dependabot
+  module NpmAndYarn
+    class PackageName
+      DEFINITELY_TYPED_SCOPE = /types/i
+      PACKAGE_NAME_REGEX     = %r{
+          \A                                         # beginning of string
+          (?=.{1,214}\z)                             # enforce length (1 - 214)
+          (@(?<scope>[a-z0-9\-~][a-z0-9\-\._~]*)\/)? # capture 'scope' if present
+          (?<name>[a-z0-9\-~][a-z0-9\-._~]*)         # capture package name
+          \z                                         # end of string
+      }xi.freeze                                     # multi-line/case-insensitive
+
+      class InvalidPackageName < StandardError; end
+
+      def initialize(string)
+        match = PACKAGE_NAME_REGEX.match(string.to_s)
+        raise InvalidPackageName unless match
+
+        @scope = match[:scope]
+        @name  = match[:name]
+      end
+
+      def to_s
+        if scoped?
+          "@#{@scope}/#{@name}"
+        else
+          @name.to_s
+        end
+      end
+
+      def <=>(other)
+        to_s <=> other.to_s
```

Comment:
We should either be more discerning on what we accept (lowercase only), or make this comparison case insensitive.

```suggestion
        to_s.casecmp(other.to_s)
```

I slightly favor the latter, since we care more about matching declared intent more than enforcing [npm moniker rules](https://blog.npmjs.org/post/168978377570/new-package-moniker-rules.html)

Reply:
Oh, actually — that's I forgot that's what npm calls them. _moniker_. I'm fine with `PackageName` as a class name, but `PackageMoniker` would also be a reasonable alternative, I think.

Reply:
I addressed this (and your other suggestions) in [997cade](https://github.com/dependabot/dependabot-core/pull/5001/commits/997cadecd2b823f4b4c8ece5bae0a7b7c4777c9d) 🙂 

I opted to keep `PackageName` as that npm blog post used "moniker" only in the title, then used "name" throughout the contents of the post. I had also referenced a package wherein npm was validating "names". The term "name" seems less esoteric and I couldn't think of any semantic meaning that's lost, so I ran with it.